test/tools/router_check/coverage.cc

```diff
@@ -5,19 +5,86 @@
 #include "envoy/api/v2/core/base.pb.h"
 
 namespace Envoy {
-void Coverage::markCovered(const Envoy::Router::RouteEntry& route) {
-  // n.b. If we reach the end of the seen routes without finding the specified
-  // route we add it as seen, otherwise it's a duplicate.
-  if (std::find(seen_routes_.begin(), seen_routes_.end(), &route) == seen_routes_.end()) {
-    seen_routes_.push_back(&route);
+double RouteCoverage::report() {
+  uint64_t route_weight = 0;
+  for (const auto& covered_field : coverageFields()) {
+    if (covered_field) {
+      route_weight += 1;
+    }
   }
+  return static_cast<double>(route_weight) / coverageFields().size();
+}
+
+void Coverage::markClusterCovered(const Envoy::Router::RouteEntry& route) {
+  coveredRoute(route)->setClusterCovered();
+}
+
+void Coverage::markVirtualClusterCovered(const Envoy::Router::RouteEntry& route) {
+  coveredRoute(route)->setVirtualClusterCovered();
+}
+
+void Coverage::markVirtualHostCovered(const Envoy::Router::RouteEntry& route) {
+  coveredRoute(route)->setVirtualHostCovered();
+}
+
+void Coverage::markPathRewriteCovered(const Envoy::Router::RouteEntry& route) {
+  coveredRoute(route)->setPathRewriteCovered();
+}
+
+void Coverage::markHostRewriteCovered(const Envoy::Router::RouteEntry& route) {
+  coveredRoute(route)->setHostRewriteCovered();
+}
+
+void Coverage::markRedirectPathCovered(const Envoy::Router::RouteEntry& route) {
+  coveredRoute(route)->setRedirectPathCovered();
 }
 
 double Coverage::report() {
   uint64_t num_routes = 0;
   for (const auto& host : route_config_.virtual_hosts()) {
-    num_routes += host.routes_size();
+    for (const auto& route : host.routes()) {
+      if (route.route().has_weighted_clusters()) {
+        num_routes += route.route().weighted_clusters().clusters_size();
+      } else {
+        num_routes += 1;
+      }
+    }
   }
-  return 100 * static_cast<double>(seen_routes_.size()) / num_routes;
+  return 100 * static_cast<double>(covered_routes_.size()) / num_routes;
 }
+
+double Coverage::detailedReport() {
+  uint64_t num_routes = 0;
+  for (const auto& host : route_config_.virtual_hosts()) {
+    for (const auto& route : host.routes()) {
+      if (route.route().has_weighted_clusters()) {
+        num_routes += route.route().weighted_clusters().clusters_size();
+      } else {
+        num_routes += 1;
+      }
+    }
+  }
+  double cumulative_coverage = 0;
+  for (auto& covered_route : covered_routes_) {
+    cumulative_coverage += covered_route->report();
+  }
+  return 100 * cumulative_coverage / num_routes;
+}
+
+void Coverage::cleanup() {
+  for (auto& covered_route : covered_routes_) {
+    delete covered_route;
```

Comment:
Store unique_ptr in your list, and then you can remove all of this manual cleanup code. (In general for C++ if you find yourself writing `delete` something is probably wrong.)